scapy/contrib/isotp.uts

```diff
@@ -0,0 +1,841 @@
+assert(len(fragments) == 2)
+assert(fragments[0].data == b"\x10\x08abcdef")
+assert(fragments[1].data == b"\x21gh")
+
+= Fragment an ISOTP message with extended addressing
+isotp = ISOTP("abcdef", exdst=ord('A'))
+fragments = isotp.fragment()
+assert(len(fragments) == 1)
+assert(fragments[0].data == b"A\x06abcdef")
+
+= Fragment a 7 bytes ISOTP message with destination identifier
+isotp = ISOTP("abcdefg", dst=0x64f)
+fragments = isotp.fragment()
+assert(len(fragments)  == 1)
+assert(fragments[0].data == b"\x07abcdefg")
+assert(fragments[0].identifier == 0x64f)
+
+= Fragment a 16 bytes ISOTP message with extended addressing
+isotp = ISOTP("abcdefghijklmnop", dst=0x64f, exdst=ord('A'))
+fragments = isotp.fragment()
+assert(len(fragments) == 3)
+assert(fragments[0].data == b"A\x10\x10abcde")
+assert(fragments[1].data == b"A\x21fghijk")
+assert(fragments[2].data == b"A\x22lmnop")
+assert(fragments[0].identifier == 0x64f)
+assert(fragments[1].identifier == 0x64f)
+assert(fragments[2].identifier == 0x64f)
+
+= Defragment a single-frame ISOTP message
+fragments = [CAN(identifier=0x641, data=b"\x04test")]
+isotp = ISOTP.defragment(fragments)
+isotp.show()
+assert(isotp.data == b"test")
+
+= Defragment an ISOTP message composed of multiple CAN frames
+fragments = [
+    CAN(identifier=0x641, data=dhex("41 10 10 61 62 63 64 65")),
+    CAN(identifier=0x641, data=dhex("41 21 66 67 68 69 6A 6B")),
+    CAN(identifier=0x641, data=dhex("41 22 6C 6D 6E 6F 70 00"))
+]
+isotp = ISOTP.defragment(fragments)
+isotp.show()
+assert(isotp.data == dhex("61 62 63 64 65 66 67 68 69 6A 6B 6C 6D 6E 6F 70"))
+
+= Check if fragmenting a frame and defragmenting it back yields the original frame
+isotp1 = ISOTP("abcdef", exdst=ord('A'))
+fragments = isotp1.fragment()
+isotp2 = ISOTP.defragment(fragments)
+isotp2.show()
+assert(isotp1 == isotp2)
+
+isotp1 = ISOTP("abcdefghijklmnop")
+fragments = isotp1.fragment()
+isotp2 = ISOTP.defragment(fragments)
+isotp2.show()
+assert(isotp1 == isotp2)
+
+isotp1 = ISOTP("abcdefghijklmnop", exdst=ord('A'))
+fragments = isotp1.fragment()
+isotp2 = ISOTP.defragment(fragments)
+isotp2.show()
+assert(isotp1 == isotp2)
+
+
++ Testing ISOTPSniffer.ISOTPMachine
+
+= Crate ISOTP state machine object
+m = ISOTPSniffer.ISOTPMachine()
+
+= Feed packets to machine
+m.feed(CAN(identifier=0x241, data=dhex("10 28 01 02 03 04 05 06")))
+m.feed(CAN(identifier=0x641, data=dhex("30 03 00"               )))
+m.feed(CAN(identifier=0x241, data=dhex("21 07 08 09 0A 0B 0C 0D")))
+m.feed(CAN(identifier=0x241, data=dhex("22 0E 0F 10 11 12 13 14")))
+m.feed(CAN(identifier=0x241, data=dhex("23 15 16 17 18 19 1A 1B")))
+m.feed(CAN(identifier=0x641, data=dhex("30 03 00"               )))
+m.feed(CAN(identifier=0x241, data=dhex("24 1C 1D 1E 1F 20 21 22")))
+m.feed(CAN(identifier=0x241, data=dhex("25 23 24 25 26 27 28"   )))
+
+= Verify there is a ready message in the machine
+assert(m.count() == 1)
+
+= Extract the message from the machine
+msg = m.pop()
```

Comment:
add check if m.length is modified correctly after pop

Reply:
Done